index.bs

```diff
@@ -1710,3 +1738,161 @@ When [=a browsing context is discarded=], run the following steps:
 
 </div>
+
+## Log ## {#module-log}
+
+The <dfn export for=modules>log</dfn> module contains functionality and events
+related to logging.
+
+### Definition ### {#module-log-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+
+Command = (GetCommand)
+</pre>
+
+[=local end definition=]
+
+<pre class="cddl local-cddl">
+
+LogEvent = (
+  LogEntryAddedEvent
+)
+</pre>
+
+### Types ### {#module-log-types}
+
+#### Log Entry #### {#types-log-log-entry}
+
+<pre class="cddl local-cddl">
+
+LogLevel = "debug" / "info" / "warning" / "error"
+
+LogEntry = {
+  type: text,
+  level: LogLevel,
+  text: text / null,
+  timestamp: int,
+  extra: { *text => any }
+}
+
+ConsoleLogExtra = {
+  type: text,
+  realm: Realm,
+  args: [*RemoteValue],
+}
+
+JavascriptLogExtra = {
+  lineNumber: int,
+  columnNumber: int,
+  url: text
+}
+
+</pre>
+
+Each log event is represented by a <code>LogEntry</code> object. This has a
+<code>type</code> property which represents the type of log entry added, a
+<code>level</code> property representing severity, a <code>text</code> property
+with the log message string itself, and a <code>timesteamp</code> property
+corresponding to the time the log entry was generated. The <code>extra</code>
+property contains additional data specific to the given type.
+
+#### entryAdded #### {#event-log-entryAdded}
+
+<dl>
+   <dt>Event Type</dt>
+   <dd>
+      <pre class="cddl local-cddl">
+        LogEntryAddedEvent = {
+         method: "log.entryAdded",
+         params: LogEntry
+       }
+      </pre>
+   </dd>
+</dl>
+The [=remote end event trigger=] is:
+
+<div algorithm="remote end event trigger for log.entryAdded">
+
+When the [=Printer=] algorithm is invoked with arguments |logLevel|, |args| and,
+optionally, <var ignore>options</var>:
+
+1. If |logLevel| is "<code>error</code>" or "<code>assert</code>", let |level|
+   be "<code>error</code>". If |logLevel| is "<code>debug</code>" or
+   "<code>trace</code>" let |level| be "<code>debug</code>". If |logLevel| is
+   "<code>warn</code>", let |level| be "<code>warning</code>". Otherwise let
+   |level| be "<code>info</code>".
+
+1. Let |timestamp| be a [=time value=] representing the current date and time in UTC.
+
+1. Let |text| be an implementation-defined string representing the arguments
+   passed as |args|, or null if there is no appropriate string representation.
+
+1. Let |serialized args| be a new list.
+
+1. For each |arg| of |args|, append the result of [=serialize as a remote
+   value=] given |arg| to |serialized args|.
+
+1. Let |realm| be the [=realm id=] of the [=current Realm Record=].
+
+1. Let |extra| be a map matching the <code>ConsoleLogExtra</code> production,
+   with the <code>type</code> field set to |logLevel|, the <code>real</code>
+   field set to |realm| and the <code>args</code> field set to |serialized
+   args|.
+
+1. Let |entry| be a map matching the <code>LogEntry</code> production, with the
```

Comment:
This is a lot of glue code. As we keep going, is there a path to abstracting away most of this, so that events could just return maps and something else turns those into the right productions? Something a bit handwavy like https://dom.spec.whatwg.org/#concept-event-create which nonetheless everyone knows how to interpret.

Reply:
So there's no difference between a map and a map that matches a production other that the latter is guaranteed to match a production. I was thinking we could have some pseudo-varargs function like:

When required to <dfn>make a map matching CDDL production</dfn> |production| with a variable number of arguments |args|:

1. Let |result| be an [=map=] matching the production |production|, specifying the fields according to the values of |args| as follows:
  * If |arg| is not a [=tuple=], set the field with the name of |arg| to the value of arg.
  * If |arg| is a tuple, set the field named by the first item of |arg| to the second item of |arg|.

1. Assert: |result| matches |production|

1. Return |result|

And then you would use it like:

1. [=Make a map matching the CDDL production=] <code>LogEntry</code> with ("type", "console"), |level|, |text|, |timestamp| and |extra|/